### ISSUE TITLE
bigbick123456789000 - Unauthorized Transfers of ETH from OptimismPortal2 Due to Premature Finalization of Withdrawals Before L2 Blockchain Confirmation

### ISSUE THREAD

**Issue by sherlock-admin3** [sherlock-audit/2024-02-optimism-2024-judging#45]
bigbick123456789000

high

# Unauthorized Transfers of ETH from OptimismPortal2 Due to Premature Finalization of Withdrawals Before L2 Blockchain Confirmation

## Summary
The vulnerability in the `OptimismPortal2` contract allows users to finalize withdrawals without verifying whether the withdrawal transactions have been included in the state of the `L2ToL1MessagePasser` contract on the Layer 2 blockchain, potentially enabling unauthorized transfers of ETH out of the contract. This lack of proper validation violates the security requirement specified in the protocol documentation, which aims to ensure that transfers are only triggered during the finalization of valid messages sent from the L2 blockchain.
## Vulnerability Detail
The contract allows for withdrawals to be immediately marked as finalized before they are processed on the L2 blockchain. This is evident from the line:
```solidity
finalizedWithdrawals[withdrawalHash] = true;
```
This means that once a withdrawal is proven, it can be marked as finalized without waiting for confirmation from the L2 blockchain.
The contract transfers ETH to the target contract (`_tx.target`) before resetting the `l2Sender` back to its default value. This is evident from the lines:
```solidity
bool success = SafeCall.callWithMinGas(_tx.target, _tx.gasLimit, _tx.value, _tx.data);
// Reset the l2Sender back to the default value.
l2Sender = Constants.DEFAULT_L2_SENDER;
```
This means that even if the withdrawal is not processed on the L2 blockchain, the ETH transfer to the target contract occurs before `l2Sender` is reset, potentially allowing unauthorized transfers of ETH.
Combining these creates a scenario where a user can potentially exploit the contract as follows:
- `Initiation of Withdrawal`: An attacker initiates a withdrawal transaction and provides a proof of withdrawal.
- `Immediate Finalization`: The contract immediately marks the withdrawal as finalized upon submission of the proof, without waiting for confirmation from the L2 blockchain.
- `ETH Transfer`: The contract transfers ETH to the target contract without resetting l2Sender.
- `Potential Exploitation`: The attacker exploits the window between finalization on L1 and processing on L2 to manipulate the process and withdraw ETH that they are not entitled to. Since the withdrawal has been marked as finalized, the attacker can potentially manipulate the state of the contract to their advantage before l2Sender is reset.
This vulnerability poses a significant risk as it allows users to exploit the timing mismatch between finalization on L1 and processing on L2, potentially leading to unauthorized transfers of ETH out of the `OptimismPortal2` contract.
## Impact
This vulnerability poses a significant risk as it allows attackers to exploit the timing mismatch between finalization on L1 and processing on L2, potentially leading to unauthorized transfers of ETH out of the `OptimismPortal2` contract. It undermines the integrity of the withdrawal process and exposes the contract to financial loss.
## Code Snippet
[OptimismPortal2.sol#L329-L331](https://github.com/sherlock-audit/2024-02-optimism-2024/blob/main/optimism/packages/contracts-bedrock/src/L1/OptimismPortal2.sol#L329-L331)
[OptimismPortal2.sol#L336-L384](https://github.com/sherlock-audit/2024-02-optimism-2024/blob/main/optimism/packages/contracts-bedrock/src/L1/OptimismPortal2.sol#L336-L384)
## Tool used

Manual Review

## Recommendation
Ensure that withdrawals are finalized only after they have been confirmed and processed on the L2 blockchain. Additionally, the sequence of operations should be carefully managed to prevent unauthorized transfers of ETH. Measures such as resetting `l2Sender` before any ETH transfer can help prevent exploitation of the vulnerability.

Duplicate of #145
